docs/proposals/diagnose-experience.md

```diff
@@ -0,0 +1,250 @@
+
+- **Purpose**: Debug logs are specifically aimed to facilitate ORAS developers to diagnose ORAS tool itself. They contain detailed technical information that is useful for troubleshooting problems.
+- **Target users**: Primarily intended for developers or technical users who are trying to understand the inner workings of the code and identify the root cause of a possible issue with the tool itself.
+- **Content**: Debug logs focus on providing context needed to troubleshoot issues, like variable values, execution paths, error stack traces, and internal states of the application.
+- **Level of Detail**: Extremely detailed, providing insights into the application's internal workings and logic, often including low-level details that are essential for debugging.
+
+### Common Conventions
+
+Here are the common conventions to print clear and analyzable debug logs.
+
+### **Timestamp Each Log Entry**
+- **Precise Timing:** Ensure each log entry has a precise timestamp to trace the sequence of events accurately.
+  - Example: `DEBUG: [2023-10-01T12:00:00Z] Starting metadata retrieval for repository oras-demo`
+
+### **Avoid Logging Sensitive Information**
+- **Privacy and Security:** Abstain from logging sensitive information such as passwords, personal data, or authentication tokens.
+  - Example: `DEBUG: Attempting to authenticate user [UserID: usr123]`  (exclude authentication token and password information).
+
+## Proposals for ORAS CLI
+
+Based on the concepts and conventions above, here are the proposal for ORAS diagnose experience improvement: 
+
+- Deprecate the `--verbose` flag and keep `--debug` flag to avoid ambiguity. It is reasonable to continue using `--debug` to enable the output of `DEBUG` level logs as it is in ORAS. Meanwhile, This change will make the diagnose experience much more straightforward and less breaking since only ORAS `pull/push/attach/discover` commands have verbose output.
+- Make the verbose output of commands `pull`, `push`, `attach` as the default (status) output. See examples at the bottom.
+- Make the verbose output of command  `discover` as a formatted output, controlled by `--format tree-full`. 
+- Add two empty lines as the separator between each request and response for readability.
+- Add timestamp of each request and response to the beginning of each request and response.
+- Upon success, print out the response body in the debug logs if the `mediaType` of the processed OCI object (e.g. image manifest) is JSON or plain text format. The [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) should be included in the response body by default for diagnose purposes.
```

Comment:
Considering the security, what's the size limit of printing the response body?

Reply:
@shizhMSFT 

Size limit is a valid concern as it is crucial for preventing ORAS clients from overloading the registry server with massive payloads that could lead to performance issues. 

However, I don't have a certain value for the size limit on the ORAS client side. Is it depends on the OCI registry? According to the [OCI Distribution Spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pushing-manifests), should ORAS determines the maximum size of the response body based on the specific registry's implementation?

> A registry SHOULD enforce some limit on the maximum manifest size that it can accept. A registry that enforces this limit SHOULD respond to a request to push a manifest over this limit with a response code 413 Payload Too Large.

Reply:
We can decide a value by ourselves.

Reply:
Revised this sentence accordingly.